rfcs/syntax-string-interpolation.md

```diff
@@ -0,0 +1,141 @@
+# String interpolation
+
+## Summary
+
+New string interpolation syntax.
+
+## Motivation
+
+The problems with `string.format` are many.
+
+1. Must be exact about the types and its corresponding value.
+2. Using `%d` is the idiomatic default for most people, but this loses precision.
+   * `%d` casts the number into `long long`, which has a lower max value than `double` and does not support decimals.
+   * `%f` by default will format to the millionths, e.g. `5.5` is `5.500000`.
+   * `%g` by default will format up to the hundred thousandths, e.g. `5.5` is `5.5` and `5.5312389` is `5.53123`. It will also convert the number to scientific notation when it encounters a number equal to or greater than 10^6.
+   * To not lose too much precision, you need to use `%s`, but even so the type checker assumes you actually wanted strings.
+3. No support for `boolean`. You must use `%s` **and** call `tostring`.
+4. No support for values implementing the `__tostring` metamethod.
+5. Using `%` is in itself a dangerous operation within `string.format`.
+   * `"Your health is %d% so you need to heal up."` causes a runtime error because `% so` is actually parsed as `(%s)o` and now requires a corresponding string.
+6. Having to use parentheses around string literals just to call a method of it.
+
+## Design
+
+To fix all of those issues, we need to do a few things.
+
+1. A new string interpolation expression (fixes #5, #6)
+2. Extend `string.format` to accept values of arbitrary types (fixes #1, #2, #3, #4)
+
+Because we care about backward compatibility, we need some new syntax in order to not change the meaning of existing strings. There are a few components of this new expression:
+
+1. A string chunk (`` `...{ ``, `}...{`, and `` }...` ``) where `...` is a range of 0 to many characters.
+   * `%` escapes `%` and `{`, whereas `\` escapes `` ` ``.
```

Comment:
What's the rationale for treating `%` as a special character within the interpolated strings? I know we're lowering to string.format, but it looks like `%` isn't a special character other than being used for escapes. If we use `\` to escape `{`, the syntax will be simpler. We'd need to automatically escape `%` in the lowering process but the user doesn't need to know.

Reply:
Yeah the rationale was simply about using `%` as the character that escapes something in a format string, mainly to be consistent with `string.format` itself. `\` is definitely a viable option to escape opening brace or itself. Would certainly simplify the internal parsing code a bit.

Reply:
Changed it to backslashes.